src/components/document/Document.js

```diff
@@ -194,60 +300,125 @@ function Document({ docUrl, data, loading, fullPage, showToolbar = true, showMet
+                )}
 
-            <TextSection
-                title="Headers &amp; Parts"
-                text={data.content.tree}
-                fullPage={fullPage}
-            />
+                {ocrData.map(({tag}) => (
+                    <TabStyle icon={<TextFields />} label={"OCR " + tag} classes={tabClasses} key={tag} />
+                ))}
+
+                {!!data.children?.length && (
+                    <TabStyle icon={<FolderOutlined />} label="Files" classes={tabClasses} />
+                )}
+            </Tabs>
+
+            {!!data.content.text && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <Text content={data.content.text} />
+                </TabPanel>
+            )}
+
+            {hasPreview && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <Preview
+                        docTitle={data.content.filename}
+                        type={data.content["content-type"]}
+                        url={docRawUrl}
+                    />
+                </TabPanel>
+            )}
+
+            {data.content.filetype !== 'folder' && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <Tags
+                        loading={tagsLoading}
+                        digestUrl={digestUrl}
+                        tags={tags}
+                        onChanged={setTags}
+                        toolbarButtons={tagsLinks}
+                        locked={tagsLocked}
+                        onLocked={setTagsLocked}
+                    />
+                </TabPanel>
+            )}
+
+            {!!data.content && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <Meta
+                        doc={data}
+                        collection={collection}
+                        baseUrl={collectionBaseUrl}
+                    />
+                </TabPanel>
+            )}
+
+            {data.content.filetype === 'email' && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <Email doc={data} collection={collection} />
+                </TabPanel>
+            )}
+
+            {!!data.safe_html && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <HTML html={data.safe_html} />
+                </TabPanel>
+            )}
+
+            {!!data.content.tree && (
+                <TabPanel value={tab} index={tabIndex++}>
+                    <Text content={data.content.tree} />
+                </TabPanel>
+            )}
 
             {ocrData.map(({tag, text}) => (
-                <TextSection
-                    key={tag}
-                    title={"OCR " + tag}
-                    text={text}
-                    fullPage={fullPage}
-                    omitIfEmpty={false}
-                />
+                <TabPanel value={tab} index={tabIndex++} key={tag}>
```

Comment:
what's up with the weird duplication of all these fields here and on line 339?

We want to be moving towards field definitions that come from the backend, not repeated hand-rolled loops...

Maybe some OOP / abstraction would help here?

Reply:
added configuration